docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,500 @@
+# Stateless Ceph-CSI plugin implementation
+
+## Terminology
+
+- CO: Container orchestration system, communicates with CSI plugins using CSI
+  RPCs
+- NS: Node Service; CSI plugin service per node in the CO, responsible for
+  staging volumes for application pod consumption
+- CS: Controller Service; CSI plugin service responsible for creation,
+  deletion, snapshots, and expanding volumes for use in the CO system
+
+**NOTE:**
+
+- Discussion below may use k8s specific terminology, but should be extensible
+  to similar constructs on other container orchestration(CO) systems
+
+## Proposal/problem summary
+
+This document details a proposed solution, among presented alternatives, to
+implement a stateless Ceph-CSI plugin. It also details a naming scheme for
+volumes provisioned against a Ceph cluster and the resultant *volume_id* for
+the same that is shared with the CO system, enabling state to be carried by the
+*volume_id* where required.
+
+As of this writing, the plugin uses a config map to store *volume_name* to
+*volume_id* mapping, and also stores related Ceph cluster configuration
+information (e.g MON list, pool information) within the same tuple, to be able
+to find and operate on the provisioned volume. The implementation also depends
+on secrets from the StorageClass to perform NS related operations.
+
+The existing strategy enables a single CSI instance to operate across any Ceph
+cluster that is fully defined by the StorageClass, but brings in the burden of
+storing and managing provisioned volumes by the plugin.
+
+The proposal laid out here is to try and retain the ability for a single CSI
```

Comment:
Do we need multiple replicas? If you had multiple, how would they coordinate the requests? Provisioning with ceph with the flex driver has been pretty quick in rook so if we don't need multiple the simplification is always good.

Reply:
Removed multiple controller service possibility from the text.